### ISSUE TITLE
AvroJSONDecoder returns error for JSON with newlines

### ISSUE THREAD

**Issue by xxchan** [fastavro/fastavro#778]
It seems it expects one line is a value, which seems a little strange.

https://github.com/fastavro/fastavro/blob/bec17a72471434f12cec04c5575ed654ade345b9/fastavro/io/json_decoder.py#L47C14-L47C24

**Comment by scottbelden** [comment]
Yes, this is true and was done because it was easy. As far as I know the standard python json library doesn't have a way to parse an arbitrary input into tokens.

**Comment by shoffmeister** [comment]
@scottbelden Would it be an option to "open up" `AvroJSONDecoder` a little bit?

Right now the class documentation says
```
    NOTE: All attributes and methods on this class should be considered
    private.
```

Case in point: By introducing some method such as `ingest_all_json_data()` (below, untested), the Avro JSON content splitting logic could be split out.

It would then be easy to introduce a custom decoder (such as the `CustomJSONDecode` in the tests), which implements its own splitting logic - provided that the stern notice about "private" gets softened up a bit.

My local use case is, for instance, reading a single pretty-printed Avro JSON. I solved that for now by simply copying `__init__()` at large, but that is Not At All Nice.

Happy to provide a PR with `self.ingest_all_json_data(self, fo)` or whatever naming you feel most appropriate.

```
    def __init__(self, fo: IO):
        self._fo = fo
        self._stack: List[Tuple[Any, str]] = []
        self._json_data = self.ingest_all_json_data(self, fo)
        if self._json_data:
            self._current = self._json_data.pop(0)
            self.done = False
        else:
            self.done = True
        self._key = None

    def ingest_all_json_data(self, fo: IO) -> list[Any]:
        return [json.loads(line.strip()) for line in fo]
```
